### ISSUE TITLE
Change Request: Spin out lib/unsupported-features as a standalone package?

### ISSUE THREAD

**Issue by JoshuaKGoldberg** [eslint-community/eslint-plugin-n#511]
### eslint-plugin-n version

v17.23.2

### What problem do you want to solve?

[`lib/unsupported-features`](https://github.com/eslint-community/eslint-plugin-n/tree/7541d64f4748b7c0f9cd945d8dccd78cdda15464/lib/unsupported-features) is a treasure trove of info about which Node.js features were added in which Node.js versions. I'd like to use it in other tools that check for Node.js feature <> version compatibility. For example, the experimental https://github.com/flint-fyi/flint/issues/932. And I'd be surprised if other linters such as Biome and Oxlint don't also end up including equivalents to `n/no-unsupported-features/*`.

But, `lib/unsupported-features` right now seems to be an internal implementation detail of the `eslint-plugin-n` package. There are no docs for it and the package doesn't export it in `package.json` or `lib/index.js`. There's no good, sanctioned way for us to intake that data that I can find.

### What do you think is the correct solution?

Would the `eslint-plugin-n` team be open to publishing the data under `lib/unsupported-features` as a standalone package? Like a `node-feature-support` package or some similar name?

I imagine the two most common solutions to do so would be either:

* Switching this repository to a monorepo
* Creating a separate repository for the new package

### Participation

- [x] I am willing to submit a pull request for this change.

### Additional comments

I'd of course be happy to help maintain this if you're up for it and would find my help useful.

If this request is declined (totally understandable) I'll likely end up making my own library for this.

Thanks! ❤️

**Comment by scagood** [comment]
I love this idea.

Of the two options, I would be in favour of a monorepo.

**Comment by JoshuaKGoldberg** [comment]
> I'm not sure if these symbols are that specific to ESLint tbh

I can't find any docs on them - why do they exist in the first place? Why not go with plain keys like `{ read: ... }`? 

From a medium-level inspection, the usage of symbols looks like an eslint-community-specific implementation detail that other consumers almost certainly won't want to take on. I see no reason for them in an arbitrary other linter. Unless I'm missing something?

> `READ`/`CALL`/`CONSTRUCT`

I also don't follow why there needs to be a delineation between them. Every `CALL` and `CONSTRUCT` usage in userland code is, by definition, also a `READ`. You can't call or construct something without a reference to it! Are there some APIs that switched their read/call/construct characteristics between Node.js versions? Am I missing something here too?

**Comment by aladdin-add** [comment]
> Will we release it in sync with and with same version number as the plugin?

I don't think so - it's just an independent dep.

**Comment by voxpelli** [comment]
> > Will we release it in sync with and with same version number as the plugin?
> 
> I don't think so - we can use the independent mode, something like: https://lerna.js.org/docs/features/version-and-publish#independent-mode

Then we would have to swap release-please for Lerna, right? That sounds like much extra work and complexity compared to just splitting

**Comment by scagood** [comment]
> I don't think so - we can use the independent mode, something like: [lerna.js.org/docs/features/version-and-publish#independent-mode](https://lerna.js.org/docs/features/version-and-publish#independent-mode)

<details>
<summary>I personally use release please to manage my monorepos</summary>

<img width="655" height="365" alt="Image" src="https://github.com/user-attachments/assets/11afa674-ef9e-4f3e-8548-4af02a628251" />

</details>

```
  "include-component-in-tag": true,
  "plugins": [ "node-workspace" ],
  "separate-pull-requests": false,
```

---

I think the reason we use symbols is because the way the [`TraceMap`](https://github.com/eslint-community/eslint-plugin-n/blob/master/lib/eslint-utils.d.ts#L36-L41) work:

```ts
type TraceMap<Info extends unknown> = {
    [READ]?: Info;
    [CALL]?: Info;
    [CONSTRUCT]?: Info;
    [key: string]: TraceMap<Info> | undefined;
}
```

That being you can have a symbol, and child properties at the same time, an example that comes to mind is [WebCrypto](https://github.com/eslint-community/eslint-plugin-n/blob/master/lib/unsupported-features/node-builtins-modules/crypto.js#L9-L10) where I mark `WebCrypto` as experimental, and then mark properties below it as supported.

---

> I also don't follow why there needs to be a delineation between them. Every CALL and CONSTRUCT usage in userland code is, by definition, also a READ. You can't call or construct something without a reference to it! Are there some APIs that switched their read/call/construct characteristics between Node.js versions? Am I missing something here too?

This is because we track deprecated and supported, eg `new Buffer()` is deprecated, but `Buffer.from` is supported ([lib/unsupported-features/node-builtins-modules/buffer.js:26](https://github.com/eslint-community/eslint-plugin-n/blob/master/lib/unsupported-features/node-builtins-modules/buffer.js#L26-L39)).

**Comment by voxpelli** [comment]
> If there is no plan or perspective that this repo will have more than the 2 packages (`eslint-plugin-n` and the new `node-feature-support`), than maybe a new repo for the new package is better to keep things simple.

+1

**Comment by morgan-coded** [comment]
Following the API-first direction from the last discussion, I can start with a written, consumer-neutral schema for the current `lib/unsupported-features` data.

That first slice would include an adapter that reproduces the current `ReferenceTracker` trace maps while keeping every existing rule result unchanged, plus golden parity tests over every current entry.

The contract would make `READ`, `CALL`, and `CONSTRUCT` explicit, including APIs deprecated only in particular modes and nested properties with their own support ranges; I’d use #359 as design input for cases the current model does not express clearly.

Repository placement, release and versioning topology, package publication, and new data sources would stay out of this pass and remain the maintainers’ call.

If that boundary fits the direction here, I can write up the schema and adapter contract for review before touching extraction or code.

**Comment by morgan-coded** [comment]
@MichaelDeBoey, one concrete inventory from `master` makes the API gap clearer.

The 1,379 declared access records expand through aliases and spreads to 3,272 consumer-visible path/access records: 3,221 `READ`, 42 `CALL`, and 9 `CONSTRUCT`. Of those expanded records, 51 deprecations differ from another access mode at the same path, and 1,129 are nested below the first exported member with their own lifecycle ranges. The current data cannot answer #359's key count, how many `READ` records name functions, because it stores lifecycle fields but no value kind.

<details>
<summary>Expanded records by defining file</summary>

```text
assert 132; async_hooks 20; buffer 52; child_process 18; cluster 26
console 48; crypto 238; dgram 6; diagnostics_channel 16; dns 142
domain 3; events 80; fs 360; http 56; http2 40; https 14
inspector 68; module 70; net 34; os 50; path 150; perf_hooks 84
process 218; punycode 7; querystring 14; readline 34; repl 22
sea 11; sqlite 12; stream 118; string_decoder 4; test 44
timers 42; tls 40; trace_events 6; tty 8; url 28; util 281
v8 64; vm 25; wasi 3; worker_threads 50; zlib 235
node-globals 295; node-import-meta 4; types 0
```
</details>

I'd use plain string keys at the consumer-neutral boundary and translate them to `ReferenceTracker` symbols only in the ESLint adapter:

```js
[
  {
    path: ["crypto", "Hash"],
    kind: "class",
    access: {
      read: { supported: ["0.1.92"] },
      call: { deprecated: ["22.0.0", "20.13.0"] },
      construct: { deprecated: ["22.0.0", "20.13.0"] },
    },
  },
  {
    path: ["process", "features", "typescript"],
    kind: "property",
    access: {
      read: {
        experimental: ["23.0.0", "22.10.0"],
        supported: ["24.12.0"],
      },
    },
  },
]
```

That separates value kind from access mode, makes #359 measurable, and preserves current symbol-backed consumers. I can still write the schema and parity contract first; repository placement, publication, versioning, and ownership remain the maintainers' call. If this seam is no longer useful, say so and I'll drop it.
